### ISSUE TITLE
Excessive time to load font with pygame.font.SysFont on Mac

### ISSUE THREAD

**Issue by IrvKalb** [pygame/pygame#2553]
- **Operating system**: Mac OSX 11.2.3
- **Python version** : 3.9.1
- **SDL version**: 2.0.14
- **PyGame version**: 2.0.1

**Current behavior:**

The first font loaded by name using pygame.font.SysFont(), is taking over 7 seconds on my Mac to load.  

**Expected behavior:**

I would expect that a font loaded by name would be almost instantaneous - as it is when using pygame.font.font()

**Steps to reproduce:**

See sample program below.  The first font loaded using a call to pygame.font.SysFont() takes a a long time.  After the first font, additional calls to pygame.font.SysFont() are almost instanteous.  Sample output:

Loading System font with pygame.font.Font took 0.0007171630859375
Loading Arial with pygame.font.SysFont took 7.108750104904175
Loading Courier with pygame.font.SysFont took 0.0005779266357421875

But the problem appears to be with pygame.font.get_fonts().  If I add in a call to that after the first call load the System font, I get this result:

Loading System font with pygame.font.Font took 0.0006759166717529297
Getting fonts with pygame.font.get_fonts() took 7.203648805618286
Loading Arial with pygame.font.SysFont took 0.000492095947265625
Loading Courier with pygame.font.SysFont took 0.0006482601165771484

This does not seem to be a problem on Windows, only Mac

**Test code**
```
import sys
import pygame
import time
 
pygame.init()
screen = pygame.display.set_mode((400, 400))
clock = pygame.time.Clock()

now = time.time()
font1 = pygame.font.Font(None, 24)
print('Loading System font with pygame.font.Font took', time.time() - now)

#Uncomment these lines to see that the slowdown is related to getting a list of fonts.
#now = time.time()
#fontList = pygame.font.get_fonts()
#print('Getting fonts with pygame.font.get_fonts() took', time.time() - now)

now = time.time()
font2 = pygame.font.SysFont('Arial', 24)
print('Loading Arial with pygame.font.SysFont took', time.time() - now)

now = time.time()
font3 = pygame.font.SysFont('Courier', 24)
print('Loading Courier with pygame.font.SysFont took', time.time() - now)


while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
     
    screen.fill((255, 255, 255))
     
    pygame.display.flip()
    clock.tick(30)
```

**Comment by Starbuck5** [comment]
Thanks for the report.

I can't seem to replicate this on my mac.

```
Loading System font with pygame.font.Font took 0.001895904541015625
Loading Arial with pygame.font.SysFont took 0.19544196128845215
Loading Courier with pygame.font.SysFont took 0.0004773139953613281
```

```
Loading System font with pygame.font.Font took 0.0006799697875976562
Getting fonts with pygame.font.get_fonts() took 0.04025697708129883
Loading Arial with pygame.font.SysFont took 0.00045680999755859375
Loading Courier with pygame.font.SysFont took 0.0003521442413330078
```

If you're up for it, you could debug this inside pygame itself in `src_py/sysfont.py`, which is easier to debug than most pygame code, because it is written in python. 

https://github.com/pygame/pygame/blob/6746053d1cda4c194a617e696b6eca06414e80b7/src_py/sysfont.py#L418
calls

https://github.com/pygame/pygame/blob/6746053d1cda4c194a617e696b6eca06414e80b7/src_py/sysfont.py#L193
which most likely calls

https://github.com/pygame/pygame/blob/6746053d1cda4c194a617e696b6eca06414e80b7/src_py/sysfont.py#L177

So if you want to debug, you could put your timers inside sysfont.py itself by finding the file in your locally installed pygame.

**Comment by IrvKalb** [comment]
I'm happy to help out.

I added timers in all the places you suggested.  And it definitely is in the lowest level call.  I added two timers in that function to see what was happening.  Here's the code with my timers (with some odd formatting):

def _system_profiler_darwin():
    now = time.time()
    fonts = {}
    flout, _ = subprocess.Popen(
        ' '.join(['system_profiler', '-xml', 'SPFontsDataType']),
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        close_fds=True
    ).communicate()
    print('Inside _system_profiler_darwin, after flout took', time.time() - now)

    now = time.time()
    for font_node in ET.fromstring(flout).iterfind('./array/dict/array/dict'):
        _add_font_paths(font_node.iter("*"), fonts)
    print('Inside _system_profiler_darwin, after bottom for loop took', time.time() - now)
    return fonts

And here are there results:

Loading System font with pygame.font.Font took 0.0007131099700927734
Inside _system_profiler_darwin, after flout took 6.852806806564331
Inside _system_profiler_darwin, after bottom for loop took 0.07542920112609863
Inside initsysfonts_darwin took 7.026477098464966
Loading Arial with pygame.font.SysFont took 7.046849012374878
Loading Courier with pygame.font.SysFont took 0.0004818439483642578

So, it's definitely in the line that starts with:

    flout, _ = subprocess.Popen(  ...

Please let me know if there is anything else I can do to help track this down.

Irv

**Comment by Starbuck5** [comment]
Maybe this is a MacOS 11 thing...

Does it work besides the delay? Like is the font you get at the end the right one?

What is the length of the list returned by `pygame.font.get_fonts()`?

Does `system_profiler -xml SPFontsDataType` work in your terminal? For me, it takes several seconds in the terminal, but is fast when in pygame.

These are just some random questions... Like maybe MacOS 11 has a bunch more fonts bundled than earlier versions, so it's having trouble passing the amount of data into python quickly...

I don't know if this will be able to be fixed.

**Comment by IrvKalb** [comment]
I know very little about typography, but I believe that after the delay, it's loading and then showing the proper font(s).

My system shows 462 fonts.  I have not added any personally, but there might have been a small number added with different projects I've done over the years.  I found a page that shows all the fonts available with Mac OS X 11 -it's a  large list:

https://support.apple.com/en-us/HT211240

I ran the system_profiler, and just counting in my head, it took about 8 to 9 seconds to respond with a big xml list.

**Comment by futurelucas4502** [comment]
I can confirm I am also getting this issue on:
macOS 11.1
Python 3.9.1
SDL 2.0.14
pygame 2.0.1

The website linked in the previous comment has 264 fonts
My local machine when running pygame.font.get_fonts() returns 244 fonts with a long delay
`system_profiler -xml SPFontsDataType` returns a plist with 249 fonts with a long delay

So ignoring the website there is a discrepancy between the number of fonts on my machine and the number pygame returns however both pygame and the `system_profiler` command take about 8 - 9 seconds

**Comment by futurelucas4502** [comment]
Also which might help out this is also an issue on macOS 10.1.5

**Comment by simonsolnes** [comment]
Can confirm that it takes a bit under 10s to load fonts under macOS 11.3.1 (Latest)

**Comment by Starbuck5** [comment]
Alright, we've gathered a small community of macOS SysFont users here.

I finally figured out why I couldn't replicate this, while debugging another issue. My Mac wasn't going down the same path.

If you check out this function in `sysfont.py`, you can see that if the system has X11, it can use that utilities for that, which must be faster. When I changed this to use the `_system_profiler_darwin()` path, mine was slow as well. Meaning that it's `system_profiler` that's slow, and it's always been slow.
```py
def initsysfonts_darwin():
    """ Read the fonts on MacOS, and OS X.
    """
    # if the X11 binary exists... try and use that.
    #  Not likely to be there on pre 10.4.x ... or MacOS 10.10+
    if exists('/usr/X11/bin/fc-list'):
        fonts = initsysfonts_unix('/usr/X11/bin/fc-list')
    # This fc-list path will work with the X11 from the OS X 10.3 installation
    # disc
    elif exists('/usr/X11R6/bin/fc-list'):
        fonts = initsysfonts_unix('/usr/X11R6/bin/fc-list')
    elif exists('/usr/sbin/system_profiler'):
        try:
            fonts = _system_profiler_darwin()
        except (OSError, ValueError):
            fonts = {}
    else:
        fonts = {}

    return fonts
```

What do you think about replacing the call to system_profiler with a search of all the font directories of macOS? I found a list of possible locations here: https://creativepro.com/font-fatigue-font-locations-mac-os-x/

If I'm not overlooking any roadblocks, that would fix this issue **and** it would also fix  #2549.

**Comment by IrvKalb** [comment]
The article you referenced may be a little old, as it has a number of references to the mac Classic environment (meaning OS 9 and earlier).

But I'm in definitely in favor of any solution that brings the font loading time down to something reasonable.  Sounds like an approach that is worth testing out.

Thanks!

**Comment by Starbuck5** [comment]
Alright y'all, I have a PR in that fixes it.

The funny thing is that I spent most of my time fixing something in `SysFont` that nobody had filed a bug report over. :smile:

If you need your SysFont fix ASAP you can clone my branch and build pygame from source, which is actually fairly easy on Mac. (At least compared to building it on Windows)

I'm confident my fix will work, but I did develop it in MacOS 10.12, so hopefully most everything in the same.

I personally recommend downloading a `ttf` file and creating Font objects without relying on the system, because it's more customizable and more portable.

**Comment by Starbuck5** [comment]
Closed by #2594

**Comment by IrvKalb** [comment]
I look forward to seeing this fix in the next release of pygame.

Thank you!

**Comment by gsugiart** [comment]
@Starbuck5 thanks for fixing this font bug. May I know in which pygame release this bug is fixed? And while waiting for the bug fix release, is there a work around?

```self.score_font = pygame.font.SysFont("Arial", 30)```

**Comment by Starbuck5** [comment]
@gsugiart The fix is in every pygame release since May, 2.0.2.dev2, 2.0.2.dev4, and 2.0.2.

**Comment by IrvKalb** [comment]
Just downloaded pygame 2.0.2 and tested it.  The fix works great!

Thank you very much
